linux/alsa/alsa_driver.c

```diff
@@ -2062,6 +2062,97 @@ discover_alsa_using_apps ()
         }
 }
 
+static int
+alsa_driver_open (alsa_driver_t *driver, bool is_capture)
```

Comment:
Indirectly relying on `SND_PCM_STREAM_CAPTURE` being 0 in `snd_pcm_stream_t` is a little sus, even if it is unlikely to actually change (considering how unpopular it would be if the ALSA API did that sort of thing). What I mean is that the `is_capture` function parameter is unnecessary and actually complicates the whole function.

How about:

```c
static int
alsa_driver_open (alsa_driver_t *driver, snd_pcm_stream_t stream)
{
	snd_pcm_t **phandle;
	const char *name;

	if (stream == SND_PCM_STREAM_CAPTURE) {
		phandle = &driver->capture_handle;
		name = driver->alsa_name_capture;
	} else {
		phandle = &driver->playback_handle;
		name = driver->alsa_name_playback;
	}

	err = snd_pcm_open (phandle, name, stream, SND_PCM_NONBLOCK);

	if (err < 0) {
		switch (errno) {
		case EBUSY:
#ifdef __ANDROID__
			jack_error ("\n\nATTENTION: The device \"%s\" is "
				    "already in use. Please stop the"
				    " application using it and "
				    "run JACK again", name);
#else
			current_apps = discover_alsa_using_apps ();
			if (current_apps) {
				jack_error ("\n\nATTENTION: The device \"%s\" is "
					    "already in use. The following applications "
					    " are using your soundcard(s) so you should "
					    " check them and stop them as necessary before "
					    " trying to start JACK again:\n\n%s", name,
					    current_apps);
				free (current_apps);
			} else {
				jack_error ("\n\nATTENTION: The device \"%s\" is "
					    "already in use. Please stop the"
					    " application using it and "
					    "run JACK again", name);
			}
#endif
			break;

		case EPERM:
			jack_error ("you do not have permission to open "
				    "the audio device \"%s\" for playback", name);
			break;

		case EINVAL:
			jack_error ("the state of handle or the mode is invalid "
				    "or invalid state change occured \"%s\" for playback", name);
			break;

		case ENOENT:
			jack_error ("device \"%s\"  does not exist for playback", name);
			break;

		case ENOMEM:
			jack_error ("Not enough memory available for allocation for \"%s\" for playback",
				    name);
			break;

		case SND_ERROR_INCOMPATIBLE_VERSION:
			jack_error ("Version mismatch \"%s\" for playback", name);
			break;
		}

		alsa_driver_delete (driver);
		*phandle = NULL;
	}

	if (*phandle)
		snd_pcm_nonblock (*phandle, 0);

	return err;
}
```